creator-node/src/snapbackSM/snapbackSM.js

```diff
@@ -1194,35 +1195,44 @@ class SnapbackSM {
    * @notice this will issue sync to healthy secondary and update replica set away from unhealthy secondary
    */
   async aggregateReconfigAndPotentialSyncOps(nodeUsers, unhealthyPeers) {
+    // Parallelize calling this._aggregateOpsWithQueriedSpIds on chunks of 500 nodeUsers at a time
+    const nodeUserBatches = _.chunk(
+      nodeUsers,
+      AGGREGATE_RECONFIG_AND_POTENTIAL_SYNC_OPS_BATCH_SIZE
+    )
+    const resultBatches = []
+    while (nodeUserBatches.length) {
```

Comment:
per our earlier discussion, i think we should swap this for a for loop
much more readable + safer - with while loop there's risk of infinite loop if internal increment logic is buggy

Reply:
I slightly disagree on the safer part because shift() is a standard function that won't have any bugs, but I can see what you're saying for readability when someone is first glancing at this. nbd -- will change this to a for loop

Reply:
I also agree with @theoilie here -- shift will pop the first ele from the arr until there's nothing left in the array. if we want to be super sure, maybe we can change the while condition to check the shift response. if shift returns something undefined, we know it's the end of the array

Reply:
shift docs:
[Return value](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/shift#return_value)
The removed element from the array; [undefined](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/undefined) if the array is empty.

Reply:
so i didn't realize what shift was doing, thanks for sharing
this is actually much worse bc we're modifying the array every single time for no reason - this is o(n) (https://stackoverflow.com/a/22615787)

more broadly, a while loop should only be used when the termination condition is not known at start time
this is a deterministic iteration, which is what for loops are for

i'm not concerned about a bug in shift, or your current code - but while loops introduce infinite loop risk and are less readable for iteration, so we should avoid

this code would be much better as a for-of (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for-await...of)

Reply:
changed to for-of!

Reply:
looking back, yeah I agree with you on the perf side of things we should iter the array with a for or while loop for perf reasons. good stuff guys!